cmd/rukpakctl/cmd/run.go

```diff
@@ -0,0 +1,182 @@
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+	"github.com/operator-framework/rukpak/internal/rukpakctl"
+	"github.com/operator-framework/rukpak/internal/util"
+)
+
+// newRunCmd creates the run command
+func newRunCmd() *cobra.Command {
+	var (
+		systemNamespace                      string
+		uploadServiceName                    string
+		caSecretName                         string
+		bundleDeploymentProvisionerClassName string
+		bundleProvisionerClassName           string
+	)
+
+	cmd := &cobra.Command{
+		Use:   "run <bundleDeploymentName> <bundleDir>",
+		Short: "Run a bundle from an upload of a local bundle directory.",
+		Long: `Run a bundle from an upload of a local bundle directory.
+
+The run subcommand allows bundle developers to quickly iterate on bundles
+they are developing, and to test how their bundle deployment pivots from
+one version to the next.
+`,
+		Example: `
+  #
+  # Initial creation of memcached-api bundle deployment:
+  #
+  $ rukpakctl run memcached-api ./memcached-api-v0.1.0/
+  bundledeployment.core.rukpak.io "memcached-api" applied
+  successfully uploaded bundle content for "memcached-api-5b9bbf8799"
+
+  #
+  # Pivot to a new bundle for the existing memcached-api bundle-deployment
+  #
+  $ rukpakctl run memcached-api ./memcached-api-v0.2.0/
+  bundledeployment.core.rukpak.io "memcached-api" applied
+  successfully uploaded bundle content for "memcached-api-8578dfddf9"
+
+  #
+  # Run the same command again
+  #
+  $ rukpakctl run memcached-api ./memcached-api-v0.2.0/
+  bundledeployment.core.rukpak.io "memcached-api" applied
+  bundle "memcached-api-8578dfddf9" is already up-to-date
+`,
+		Args: cobra.ExactArgs(2),
+		Run: func(cmd *cobra.Command, args []string) {
+			bundleDeploymentName, bundleDir := args[0], args[1]
+			ctx := signals.SetupSignalHandler()
+
+			cfg := ctrl.GetConfigOrDie()
+
+			sch := scheme.Scheme
+			if err := rukpakv1alpha1.AddToScheme(sch); err != nil {
+				log.Fatal(err)
+			}
+			cl, err := client.New(cfg, client.Options{Scheme: sch})
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			bundleFS := os.DirFS(bundleDir)
+			digest := fnv.New64a()
+			if err := util.FSToTarGZ(digest, bundleFS); err != nil {
+				log.Fatal(err)
+			}
+
+			bundleLabels := map[string]string{
+				"app":          bundleDeploymentName,
+				"bundleDigest": fmt.Sprintf("%x", digest.Sum(nil)),
+			}
+
+			bd := buildBundleDeployment(bundleDeploymentName, bundleLabels, bundleDeploymentProvisionerClassName, bundleProvisionerClassName)
+			if err := cl.Patch(ctx, bd, client.Apply, client.ForceOwnership, client.FieldOwner("rukpakctl")); err != nil {
+				log.Fatalf("apply bundle deployment: %v", err)
+			}
+			fmt.Printf("bundledeployment.core.rukpak.io %q applied\n", bundleDeploymentName)
```

Comment:
Should we add a real logger and respect logging levels set by the client? Could also be done in a follow-up.

Reply:
We could - I was trying to match `kubectl` style output as closely as I could. But definitely agree that enabling more verbosity would be a nice improvement.